ruby/ext/google/protobuf_c/defs.c

```diff
@@ -1065,6 +1072,97 @@ VALUE MessageBuilderContext_repeated(int argc, VALUE* argv, VALUE _self) {
                           name, type, number, type_class);
 }
 
+/*
+ * call-seq:
+ *     MessageBuilderContext.map(name, key_type, value_type, number,
+ *                               value_type_class = nil)
+ *
+ * Defines a new map field on this message type with the given key and value
+ * types, tag number, and type class (for message and enum value types). The key
+ * type must be :int32/:uint32/:int64/:uint64, :bool, or :string. The value type
+ * type must be a Ruby symbol (as accepted by FieldDescriptor#type=) and the
+ * type_class must be a string, if present (as accepted by
+ * FieldDescriptor#submsg_name=).
+ */
+VALUE MessageBuilderContext_map(int argc, VALUE* argv, VALUE _self) {
```

Comment:
It strikes me that this function is the equivalent of 3 or 4 lines of Ruby code, plus the special call to set the new message type as a map entry. I wonder if there is a way to implement it this way?

I envision something like:

  VALUE message_type = rb_eval(
    "if key_type == :float or key_type == :double or key_type == :enum or key_type == :message\n"
    "  raise 'Cannot add a map...'\n"
    "end\n"
    "add_message :MapEntry {\n"
    "  optional key_type, :key, 1\n"
    "  optional value_type, :value, 2\n"
    "}");
  upb_msgdef_setmapentry(ruby_to_Descriptor(message_type), true);

Don't invest too much effort if this looks tricky, but if this were possible it would make the code a lot shorter, clearer, and more robust, since it would be expressed in terms of the DSL you already previous defined.